### PR TITLE
feat: support host level dynamic setting of tls protocol version

### LINKED ISSUE #9885: feat: Support host level dynamic setting of tls protocol version
## 1. Background
In the communication between the end point product and the server, we need to consider the TLS protocol compatibility of multiple end point products. For example, some old products, old Android mobile phones, TVs and other end point devices still use the lower-level TLSv1.1 protocol version, while new products use the higher-level TLS protocol version. If the new product supports TLSv1.1, it may bring some security risks. In order to ensure that the product can establish secure communication, we need to adapt between protocol versions.

###  1.1 Problem to be solved 
At present, APISIX can only configure the version of TLS through the [config.yaml](https://github.com/apache/apisix/blob/master/conf/config-default.yaml#L106) file, and it takes effect for all hosts, which lacks flexibility. OpenResty has a feature that can specify the characteristics of the TLS protocol version according to the host in the [ssl_client_hello_by_lua_block stage](https://github.com/openresty/lua-nginx-module#ssl_client_hello_by_lua_block). APISIX can use this feature to allow users to dynamically set different TLS versions for each host according to their own needs.

### 1.2 The benefits of solving this problem
- It provides users with more flexibility to dynamically set different TLS protocols for each host according to their needs
- Enhanced APISIX dynamic configuration capabilities

## 2. Goals
In APISIX, different TLS protocol versions can be dynamically configured for each host.

## 3. Detailed design (SSL admin API)
In APISIX, a new tls protocol selection phase http_ssl_protocols_phase() is added, which is responsible for specifying the corresponding TLS protocol version in the configuration according to the SSL object matched by the SNI in the Client Hello phase of the SSL handshake.
The execution order of each stage of APISIX is as follows:

http_ssl_protocols_phase() -> http_ssl_phase() -> http_access_phase() -> ...

- `ngx_tpl.lua`
```diff
         {% if ssl.enable then %}
+        ssl_client_hello_by_lua_block {
+            apisix.http_ssl_protocols_phase()
+        }
+
         ssl_certificate_by_lua_block {
             apisix.http_ssl_phase()
         }
         {% end %}
 ```
 
### 3.1 configuration

name | Required | Type | Default | Describe
-- | -- | -- | -- | --
snis | yes |  |   | A non-empty array of HTTPS SNI
ssl_protocols | yes | string | ["TLSv1.2", "TLSv1.3"] | It is associated with snis and is used to control the SSL/TLS protocol version used between servers and clients

Simply add ssl_protocols configuration in the ssl resource, specify the protocol version of TLS for the domain name in the snis list, the default value `["TLSv1.2", "TLSv1.3"]`.
As shown in the following example, for test.com domain name, the TLS protocol version is specified as TLSv1.2, TLSv1.3.
```bash
# curl http://127.0.0.1:9180/admin/apisix/ssls/1
{
    "cert": "$cert",     
    "key": "$key",      
    "snis": ["test.com"],   
    "ssl_protocols": [
        "TLSv1.2",
        "TLSv1.3"
    ]
}
```

### 3.2 Configuration instructions
- Static configuration
The [ssl_protocols ](https://github.com/apache/apisix/blob/master/conf/config-default.yaml#L106)parameters in the static configuration will apply to all apisix nodes, but cannot be dynamically modified.
```yaml
apisix:
  ssl:
    ssl_protocols: TLSv1.2 TLSv1.3
```

- Dynamic resource allocation
Dynamic resource configuration is to create and manage ssl resources through the admin API interface of apisix. The new ssl. ssl_protocols configuration item can control fine grain for the host and dynamically specify the TLS protocol version of each host.
```bash
# curl http://127.0.0.1:9180/admin/apisix/ssls/1
{
    "cert": "$cert",     
    "key": "$key",      
    "snis": ["test.com"],   
    "ssl_protocols": [
        "TLSv1.2",
        "TLSv1.3"
    ]
}
```

The configuration will be subject to the ssl resource, and the static configuration will be overwritten . For example, if you set ssl_protocols: TLSv1.2 TLSv1.3 in config.yaml, but set ssl.ssl_protocols: [TLSv1.3] in the resource configuration, then the final apisix will use the TLSv1.3 protocol. Therefore, when using the ssl configuration of apisix, you need to pay attention to the following points:
- **SSL resource configuration will override static configuration globally, subject to resource configuration**.
- The ssl resource configuration can be dynamically modified, and the static configuration needs to be restarted apisix to take effect.
- SSL resource configuration can be controlled according to sni fine grain, static configuration only acts on apisix global.

### PR DESCRIPTION
**Pull request by AlinsRan**
### Description

Fixes #9885 

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/init.lua [hunk 1]
```diff
@@ -192,10 +211,16 @@ function _M.http_ssl_phase()
         if err then
             core.log.error("failed to fetch ssl config: ", err)
         end
+        core.log.error("failed to find any SSL certificate by SNI: ", sni)
```
Comment (leslie-tsang) [inline apisix/init.lua:214]:
```suggestion
        core.log.error("failed to match any SSL certificate by SNI: ", sni)
```

File: apisix/init.lua [hunk 2]
```diff
@@ -192,10 +211,16 @@ function _M.http_ssl_phase()
...
         ngx_exit(-1)
     end
-end
 
+    ok ,err = apisix_ssl.set_protocols_by_clienthello(ngx_ctx.matched_ssl.value.ssl_protocols)
```
Comment (leslie-tsang) [inline apisix/init.lua:218]:
```suggestion
    ok, err = apisix_ssl.set_protocols_by_clienthello(ngx_ctx.matched_ssl.value.ssl_protocols)
```

File: apisix/ssl.lua [hunk 3]
```diff
@@ -14,9 +14,10 @@
 -- See the License for the specific language governing permissions and
 -- limitations under the License.
 --
-local core = require("apisix.core")
-local ngx_ssl = require("ngx.ssl")
-local secret = require("apisix.secret")
+local core                  = require("apisix.core")
+local ngx_ssl               = require("ngx.ssl")
+local secret                = require("apisix.secret")
+local ngx_ssl_client        = require("ngx.ssl.clienthello")
```
Comment (leslie-tsang) [inline apisix/ssl.lua:20]:
```suggestion
local core           = require("apisix.core")
local secret         = require("apisix.secret")
local ngx_ssl        = require("ngx.ssl")
local ngx_ssl_client = require("ngx.ssl.clienthello")

```

File: apisix/ssl.lua [hunk 4]
```diff
@@ -38,8 +39,14 @@ local pkey_cache = core.lrucache.new {
...
 
-function _M.server_name()
-    local sni, err = ngx_ssl.server_name()
+function _M.server_name(clienthello)
+    local sni, err
+    if clienthello then
+        sni, err = ngx_ssl_client.get_client_hello_server_name()
+    else
+        sni, err = ngx_ssl.server_name()
+
```
Comment (leslie-tsang) [inline apisix/ssl.lua:48]:
```suggestion
```
Plz remove the blank line.

File: apisix/ssl.lua [hunk 5]
```diff
@@ -56,6 +63,12 @@ function _M.server_name()
     return sni
 end
 
+function _M.set_protocols_by_clienthello(ssl_protocols)
+    if ssl_protocols then
+       return ngx_ssl_client.set_protocols(ssl_protocols)
+    end
+    return true
+end
```
Comment (leslie-tsang) [inline apisix/ssl.lua:71]:
```suggestion

function _M.set_protocols_by_clienthello(ssl_protocols)
    if ssl_protocols then
       return ngx_ssl_client.set_protocols(ssl_protocols)
    end
    return true
end

```
Please follow the [code style](https://github.com/apache/apisix/blob/master/CODE_STYLE.md)

File: apisix/init.lua (hunk 2 above)
Comment (leslie-tsang) [reply apisix/init.lua:218]: BTW, I'm not sure if I should use `ngx_ctx` or `api_ctx` as a params.

File: apisix/ssl/router/radixtree_sni.lua [hunk 6]
```diff
@@ -206,13 +206,35 @@ function _M.match_and_set(api_ctx, match_only, alt_sni)
...
+
+    return true
+end
+
+
+function _M.set(matched_ssl, sni)
+    if not matched_ssl then
+        return false, "not found any matched ssl certificate"
+    end
+    local err,ok
```
Comment (leslie-tsang) [inline apisix/ssl/router/radixtree_sni.lua:228]:
```suggestion
    local ok, err
```

File: apisix/ssl/router/radixtree_sni.lua [hunk 7]
```diff
@@ -206,13 +206,35 @@ function _M.match_and_set(api_ctx, match_only, alt_sni)
...
+        return false, err
+    end
+
+    return true
+end
+
+
+function _M.set(matched_ssl, sni)
+    if not matched_ssl then
+        return false, "not found any matched ssl certificate"
```
Comment (leslie-tsang) [inline apisix/ssl/router/radixtree_sni.lua:226]:
```suggestion
        return false, "failed to match ssl certificate"
```
Would be better ?

File: docs/en/latest/ssl-protocol.md [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+`APISIX` supports set TLS protocol and also supports dynamically specifying different TLS protocol versions for each SNI.
```
Comment (leslie-tsang) [inline docs/en/latest/ssl-protocol.md:24]:
```suggestion
`APISIX` supports set TLS protocol and also supports dynamically specifying different TLS protocol versions for each [SNI](https://en.wikipedia.org/wiki/Server_Name_Indication).
```
Would be better ?

File: docs/zh/latest/ssl-protocol.md [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+`APISIX` 支持，还支持动态的为每一个 SNI 指定不同的 TLS 协议版本。
+
+**为了安全考虑，APISIX 默认使用的加密套件不支持 TLSv1.1 以及更低的版本。**
+**如果你需要启用 TLS 1.1 协议，请在 config.yaml 的配置项 apisix.ssl.ssl_ciphers 增加 TLSv1.1 协议所支持的加密套件。**
```
Comment (leslie-tsang) [inline docs/zh/latest/ssl-protocol.md:27]:
```suggestion
**为了安全考虑，APISIX 默认使用的加密套件不支持 TLSv1.1 以及更低的版本。**
**如果你需要启用 TLSv1.1 协议，请在 config.yaml 的配置项 apisix.ssl.ssl_ciphers 增加 TLSv1.1 协议所支持的加密套件。**
```

File: docs/en/latest/ssl-protocol.md [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+
+**For security reasons, the encryption suite used by default in `APISIX` does not support TLSv1.1 and lower versions.**
+**If you need to enable the TLS 1.1 protocol, please add the encryption suite supported by the TLSv1.1 protocol to the configuration item `apisix.ssl.ssl_ciphers` in `config.yaml`.**
```
Comment (leslie-tsang) [inline docs/en/latest/ssl-protocol.md:27]:
```suggestion
**For security reasons, the encryption suite used by default in `APISIX` does not support TLSv1.1 and lower versions.**
**If you need to enable the TLSv1.1 protocol, please add the encryption suite supported by the TLSv1.1 protocol to the configuration item `apisix.ssl.ssl_ciphers` in `config.yaml`.**
```

File: apisix/init.lua (hunk 2 above)
Comment (AlinsRan) [reply apisix/init.lua]: `api_ctx` has been released and should not be used.
Can only be selected in `ngx_ctx` and `local variables`.

**Comment by AlinsRan** [comment]
@kayx23 Please help review the doc.

File: docs/en/latest/ssl-protocol.md [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+    "ssl_protocols": [
+        "TLSv1.2",
+        "TLSv1.3"
+    ]
+}
+```
+
+### Notes
+
+- Dynamic configuration has a higher priority than static configuration. When the ssl_protocols configuration item in the ssl resource is not empty, the static configuration will be overridden.
```
Comment (kayx23) [inline docs/en/latest/ssl-protocol.md:61]:
```suggestion
- Dynamic configuration has a higher priority than static configuration. When the `ssl_protocols` configuration item in the ssl resource is not empty, the static configuration will be overridden.
```

File: docs/en/latest/ssl-protocol.md [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+- The static configuration applies to the entire APISIX and requires a restart of APISIX to take effect.
```
Comment (kayx23) [inline docs/en/latest/ssl-protocol.md:62]:
```suggestion
- The static configuration applies to the entire APISIX and requires a reload of APISIX to take effect.
```

File: docs/en/latest/ssl-protocol.md [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+- Dynamic configuration can control the TLS protocol version of each SNI in a fine-grained manner and can be dynamically modified, which is more flexible than static configuration.
+
+## Examples
+
+### How to specify the TLSv1.1 protocol
+
+There are some old clients that still use the lower-level TLSv1.1 protocol version, while new products use higher security-level TLS protocol versions. Enabling TLSv1.1 for new products may bring some security risks. To ensure the security of the API, we need to flexibly switch between protocol versions.
```
Comment (kayx23) [inline docs/en/latest/ssl-protocol.md:69]:
```suggestion
While newer products utilize higher security-level TLS protocol versions, there are still legacy clients that rely on the lower-level TLSv1.1 protocol. However, enabling TLSv1.1 for new products presents potential security risks. In order to maintain the security of the API, it is crucial to have the ability to seamlessly switch between different protocol versions based on specific requirements and circumstances.
```

File: docs/en/latest/ssl-protocol.md [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+For example: test.com is the domain name used by old clients and needs to be configured for TLSv1.1, while test2.com belongs to new products and supports TLSv1.2 and TLSv1.3 protocols.
+
+1. config.yaml configuration.
```
Comment (kayx23) [inline docs/en/latest/ssl-protocol.md:72]:
```suggestion
1. `config.yaml` configuration.
```

File: docs/en/latest/ssl-protocol.md [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+- Dynamic configuration has a higher priority than static configuration. When the ssl_protocols configuration item in the ssl resource is not empty, the static configuration will be overridden.
+- The static configuration applies to the entire APISIX and requires a restart of APISIX to take effect.
+- Dynamic configuration can control the TLS protocol version of each SNI in a fine-grained manner and can be dynamically modified, which is more flexible than static configuration.
+
+## Examples
+
+### How to specify the TLSv1.1 protocol
+
+There are some old clients that still use the lower-level TLSv1.1 protocol version, while new products use higher security-level TLS protocol versions. Enabling TLSv1.1 for new products may bring some security risks. To ensure the security of the API, we need to flexibly switch between protocol versions.
+For example: test.com is the domain name used by old clients and needs to be configured for TLSv1.1, while test2.com belongs to new products and supports TLSv1.2 and TLSv1.3 protocols.
```
Comment (kayx23) [inline docs/en/latest/ssl-protocol.md:70]:
```suggestion
For example, consider two domain names: `test.com`, utilized by legacy clients requiring TLSv1.1 configuration, and `test2.com`, associated with new products that support TLSv1.2 and TLSv1.3 protocols.
```

File: docs/en/latest/ssl-protocol.md [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+curl http://127.0.0.1:9180/apisix/admin/ssls/1 \
+-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+{
+     "cert" : "'"$(cat server2.crt)"'",
+     "key": "'"$(cat server2.key)"'",
+     "snis": ["test2.com"]
+}'
+```
+
+3. Access Verification
```
Comment (kayx23) [inline docs/en/latest/ssl-protocol.md:109]:
```
3. Access Verification
```
There's already a number 3 above: `3. Create an SSL object for test.com without specifying the TLS protocol version, which will use the static ...`

Same for the Chinese doc.

File: apisix/init.lua [hunk 17]
```diff
@@ -178,11 +178,30 @@ end
...
 function _M.http_ssl_phase()
+    local ok, err = router.router_ssl.set(ngx.ctx.matched_ssl)
+    if not ok then
+        if err then
+            core.log.error("failed to fetch ssl config: ", err)
+        end
+        ngx_exit(-1)
+    end
+end
+
```
Comment (leslie-tsang) [inline apisix/init.lua:189]:
```suggestion


```
Need two blank line between func.

File: docs/en/latest/ssl-protocol.md [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+apisix:
+  ssl:
+    ssl_protocols: TLSv1.2 TLSv1.3 # default TLSv1.2 TLSv1.3
+```
+
+### Dynamic Configuration
+
+Use the `ssl_protocols` field in the `ssl` resource to dynamically specify different TLS protocol versions for each SNI.
+
+* Specify the `test.com` domain uses the TLSv1.2 and TLSv1.3.
```
Comment (kayx23) [inline docs/en/latest/ssl-protocol.md:45]:
```suggestion
Specify the `test.com` domain uses the TLSv1.2 and TLSv1.3:
```

A lot of these don't need bullet points. Same for below.

File: apisix/ssl/router/radixtree_sni.lua [hunk 19]
```diff
@@ -206,13 +206,35 @@ function _M.match_and_set(api_ctx, match_only, alt_sni)
...
+
+    return true
+end
+
+
+function _M.set(matched_ssl, sni)
+    if not matched_ssl then
+        return false, "failed to match ssl certificate"
+    end
+    local ok, err
```
Comment (leslie-tsang) [inline apisix/ssl/router/radixtree_sni.lua:228]: unused var `ok` here.

File: apisix/schema_def.lua [hunk 20]
```diff
@@ -790,6 +790,15 @@ _M.ssl = {
             enum = {1, 0},
             default = 1
         },
+        ssl_protocols = {
+            description = "set ssl protocols",
+            type = "array",
+            maxItems = 3,
+            uniqueItems = true,
+            items = {
+                enum = {"TLSv1.1","TLSv1.2", "TLSv1.3"}
```
Comment (leslie-tsang) [inline apisix/schema_def.lua:799]:
```suggestion
                enum = {"TLSv1.1", "TLSv1.2", "TLSv1.3"}
```

File: apisix/ssl/router/radixtree_sni.lua (hunk 19 above)
Comment (AlinsRan) [reply apisix/ssl/router/radixtree_sni.lua:228]: Line 242 has.

Comment (AlinsRan) [reply apisix/ssl/router/radixtree_sni.lua:228]: `    ok, err = _M.set_cert_and_key(sni, new_ssl_value)`

**Review by kingluo — APPROVED** [review]

File: apisix/init.lua [hunk 21]
```diff
@@ -192,10 +212,16 @@ function _M.http_ssl_phase()
...
         ngx_exit(-1)
     end
-end
 
+    ok, err = apisix_ssl.set_protocols_by_clienthello(ngx_ctx.matched_ssl.value.ssl_protocols)
+    if not ok then
+        core.log.error("failed to set ssl protocols: ", err)
+        ngx_exit(-1)
+    end
+end
```
Comment (monkeyDluffy6017) [inline apisix/init.lua:224]: need a blank below

File: apisix/init.lua [hunk 22]
```diff
@@ -178,11 +178,31 @@ end
...
+    if not ok then
+        if err then
+            core.log.error("failed to fetch ssl config: ", err)
+        end
+        ngx_exit(-1)
+    end
+end
+
+
+function _M.http_ssl_protocols_phase()
```
Comment (monkeyDluffy6017) [inline apisix/init.lua:191]: is `http_ssl_client_hello_phase` better?

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by kingluo — APPROVED** [review]

**Pull request merged**